QLogger.cpp

```diff
@@ -1,266 +1,252 @@
-{
-   QLog_(module, LogLevel::Error, message);
-}
+Q_DECLARE_METATYPE(QLogger::LogLevel);
+Q_DECLARE_METATYPE(QLogger::LogMode);
+Q_DECLARE_METATYPE(QLogger::LogFileDisplay);
+Q_DECLARE_METATYPE(QLogger::LogMessageDisplay);
 
-void QLog_Fatal(const QString &module, const QString &message)
+namespace QLogger
 {
-   QLog_(module, LogLevel::Fatal, message);
-}
 
-void QLog_(const QString &module, LogLevel level, const QString &message)
+void QLog_(const QString &module, LogLevel level, const QString &message,
+           const QString &function, const QString &file, int line)
 {
-   const auto manager = QLoggerManager::getInstance();
-
-   QMutexLocker(&manager->mutex);
-
-   const auto logWriter = manager->getLogWriter(module);
-
-   if (logWriter && !logWriter->isStop() && logWriter->getLevel() <= level)
-   {
-      manager->writeAndDequeueMessages(module);
-      logWriter->write(module, message, level);
-   }
-   else if (!logWriter)
-      manager->queueMessage(
-          module,
-          { message, static_cast<int>(level), QDateTime::currentDateTime().toString("dd-MM-yyyy hh:mm:ss.zzz") });
+   QLoggerManager::getInstance()->enqueueMessage(module, level, message, function, file, line);
 }
 
 static const int QUEUE_LIMIT = 100;
 
-// QLoggerManager
-QLoggerManager *QLoggerManager::INSTANCE = nullptr;
-bool QLoggerManager::mIsStop = false;
-
 QLoggerManager::QLoggerManager()
-   : mutex(QMutex::Recursive)
+   : mMutex(QMutex::Recursive)
 {
-   QDir dir(QDir::currentPath());
-   dir.mkdir("logs");
+
 }
 
 QLoggerManager *QLoggerManager::getInstance()
 {
-   if (!INSTANCE)
-      INSTANCE = new QLoggerManager();
+   static QLoggerManager INSTANCE;
 
-   return INSTANCE;
+   return &INSTANCE;
 }
 
-QString QLoggerManager::levelToText(const LogLevel &level)
+bool QLoggerManager::addDestination(const QString &fileDest, const QString &module, LogLevel level,
+                                    const QString &fileFolderDestination, LogMode mode,
+                                    LogFileDisplay fileSuffixIfFull, LogMessageDisplays messageOptions, bool notify)
 {
-   switch (level)
-   {
-      case LogLevel::Trace:
-         return "Trace";
-      case LogLevel::Debug:
-         return "Debug";
-      case LogLevel::Info:
-         return "Info";
-      case LogLevel::Warning:
-         return "Warning";
-      case LogLevel::Error:
-         return "Error";
-      case LogLevel::Fatal:
-         return "Fatal";
-   }
+   QMutexLocker lock(&mMutex);
 
-   return QString();
-}
-
-bool QLoggerManager::addDestination(const QString &fileDest, const QString &module, LogLevel level)
-{
-   if (!moduleDest.contains(module))
+   if (!mModuleDest.contains(module))
    {
-      const auto log = new QLoggerWriter(fileDest, level);
+       const auto log = new QLoggerWriter((fileDest.isEmpty() ? mDefaultFileDestination : fileDest),
+                                          (level == LogLevel::Warning ? mDefaultLevel : level),
```

Comment:
I would prefer to have all this fields in a separated line. Improves readability and the cost of copying won't be incremented due to the COW property of Qt.

Reply:
Do you mean by calling the constructor and then each setters ?

Reply:
I meant to put each ternary operator into a variable and to pass each into the constructor.

Reply:
Okay, I agree ;)